src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/LocalDB.Windows.cs

```diff
@@ -0,0 +1,222 @@
+
+        // Local Db api doc https://msdn.microsoft.com/en-us/library/hh217143.aspx
+        // HRESULT LocalDBStartInstance( [Input ] PCWSTR pInstanceName, [Input ] DWORD dwFlags,[Output] LPWSTR wszSqlConnection,[Input/Output] LPDWORD lpcchSqlConnection);  
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        internal delegate int LocalDBStartInstance(
+                [In] [MarshalAs(UnmanagedType.LPWStr)] string localDBInstanceName,
+                [In]  int flags,
+                [Out] [MarshalAs(UnmanagedType.LPWStr)] StringBuilder sqlConnectionDataSource,
+                [In, Out]ref int bufferLength);
+
+        private LocalDBStartInstance localDBStartInsanceFunc = null;
+
+        private volatile SafeLibraryHandle _sqlUserInstanceLibraryHandle;
+
+        private LocalDB() { }
+
+        internal static string GetLocalDBConnectionString(string localDbInstance) =>
+            Instance.LoadUserInstanceDll() ? Instance.GetConnectionString(localDbInstance) : null;
+
+        private string GetConnectionString(string localDbInstance)
+        {
+            StringBuilder localDBConnectionString = new StringBuilder(MAX_LOCAL_DB_CONNECTION_STRING_SIZE + 1);
+            int sizeOfbuffer = localDBConnectionString.Capacity;
+            localDBStartInsanceFunc(localDbInstance, 0, localDBConnectionString, ref sizeOfbuffer);
+            return localDBConnectionString.ToString();
+        }
+
+        internal enum LocalDBErrorState
+        {
+            NO_INSTALLATION, INVALID_CONFIG, NO_SQLUSERINSTANCEDLL_PATH, INVALID_SQLUSERINSTANCEDLL_PATH, NONE
+        }
+
+        internal static uint MapLocalDBErrorStateToCode(LocalDBErrorState errorState)
+        {
+            switch (errorState)
+            {
+                case LocalDBErrorState.NO_INSTALLATION:
+                    return SNICommon.LocalDBNoInstallation;
+                case LocalDBErrorState.INVALID_CONFIG:
+                    return SNICommon.LocalDBInvalidConfig;
+                case LocalDBErrorState.NO_SQLUSERINSTANCEDLL_PATH:
+                    return SNICommon.LocalDBNoSqlUserInstanceDllPath;
+                case LocalDBErrorState.INVALID_SQLUSERINSTANCEDLL_PATH:
+                    return SNICommon.LocalDBInvalidSqlUserInstanceDllPath;
+                case LocalDBErrorState.NONE:
+                    return 0;
+                default:
+                    return SNICommon.LocalDBInvalidConfig;
+            }
+        }
+
+        /// <summary>
+        /// Loads the User Instance dll.
+        /// </summary>
+        private bool LoadUserInstanceDll()
+        {
+            // Check in a non thread-safe way if the handle is already set for performance.
+            if (_sqlUserInstanceLibraryHandle != null)
+            {
+                return true;
+            }
+
+            lock (this)
+            {
+                if(_sqlUserInstanceLibraryHandle !=null)
+                {
+                    return true;
+                }
```

Comment:
@stephentoub I tried following the recommendation in your comment. 
How the initialization is working here:
1. Try to init the library, if it fails, then error out. (assume that the failure is because LocalDB is not installed) 
2. The application continues to run and the user installs local db. 
3. Try to init the library, and it succeeds.

`LazyInitializer.EnsureInitialized` or `Lazy<T>` are good for delayed initialization. But in this case where we want delayed initialization and then re-initialization if the first initialization didn't work, there is state maintenance that is needed with the Lazy initialization alternatives. 

It does get rid of the null checks, but it doesn't lead to the reduction of any code. In fact it adds more member fields for states like, `private bool isInitialized;`

To demonstrate what I mean, I have put together a gist with the code at https://gist.github.com/saurabh500/a2de2f6da2a9bb0f706f6fc6fd12a773